discovery_child_development/utils/google_utils.py

```diff
@@ -79,3 +84,125 @@ def create_client() -> bigquery.Client:
     # Initialize a client with the provided credentials
     client = bigquery.Client(credentials=credentials)
     return client
+
+
+def write_like_condition(term: Union[str, List[str]], table: str, field: str) -> str:
+    """Create a LIKE condition for a search term or a list of search terms"""
+    if len(term) == 1:
+        return f'{table}.{field} LIKE "%{term[0]}%"'
+    else:
+        # Write an AND condition if more than one search term
+        return "(" + " AND ".join([f'{table}.{field} LIKE "%{t}%"' for t in term]) + ")"
+
+
+def create_patents_query(search_terms: List[str]) -> str:
+    """Create a query to fetch data from BigQuery using search terms:
+    the query checks search terms in the title and abstract
+
+    Args:
+    search_terms (List[str]): list of search terms
+
+    Returns
+    str: query to fetch data from BigQuery
+    """
+
+    # Create a list of 'LIKE' conditions for each search term
+    like_conditions = {}
+    for field in ["title", "abstract"]:
+        like_conditions[field] = [
+            write_like_condition(term, "gpr", field) for term in search_terms
+        ]
+
+    # Join conditions with 'OR'
+    combined_conditions_title = " OR ".join(like_conditions["title"])
+    combined_conditions_abstract = " OR ".join(like_conditions["abstract"])
+
+    q = f"""
+    WITH
+    pubs as (
+        SELECT DISTINCT
+            pub.publication_number
+        FROM `patents-public-data.patents.publications` pub
+            INNER JOIN `patents-public-data.google_patents_research.publications` gpr ON
+            pub.publication_number = gpr.publication_number
+        WHERE
+            ({combined_conditions_title})
+            OR ({combined_conditions_abstract})
+            AND pub.grant_date BETWEEN 20190101 AND 20231231
+    )
+
+    SELECT
+        gpr.publication_number,
+        url,
+        pub.grant_date,
+        title,
+        title_translated,
+        abstract,
+        abstract_translated,
+        top_terms,
+        embedding_v1,
+    FROM `patents-public-data.patents.publications` pub
+        INNER JOIN `patents-public-data.google_patents_research.publications` gpr ON
+        pub.publication_number = gpr.publication_number
+    WHERE
+        gpr.publication_number IN (SELECT publication_number FROM pubs)
+    """
+
+    return q
+
+
+def dry_run(client: bigquery.Client, query: str) -> None:
+    """Dry run a query to estimate the size of the query"""
+    job_config = bigquery.QueryJobConfig(dry_run=True, use_query_cache=False)
+    query_job = client.query(query, job_config=job_config)
+    logging.info(
+        "This query will process {} GB.".format(
+            round(query_job.total_bytes_processed / 1e9, 3)
+        )
+    )
+
+
+def upload_query_to_s3(
+    query_name: str,
+    path: str,
+    query_df: pd.DataFrame,
+    query: str,
+    metadata: List[str] = None,
+) -> None:
+    """Upload query results to S3. Saves the results in a subfolder
```

Comment:
Really good idea to timestamp the data... Should I make an issue to do the same for OpenAlex?

Reply:
Making an issue sounds like a good idea! Overall, I think lower priority than getting on with the classification algos, but can come back to it at some point.